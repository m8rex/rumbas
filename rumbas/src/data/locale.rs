use crate::data::optional_overwrite::*;
use crate::data::template::{Value, ValueType};
use schemars::JsonSchema;
use serde::Deserialize;
use serde::Serialize;

optional_overwrite! {
    pub struct Locale {
        name: String, //TODO: document names best used for shareability?
        /// The locale to use in the Numbas interface
        numbas_locale: SupportedLocale
    }
}

/// Locales supported by Numbas
#[derive(Serialize, Deserialize, JsonSchema, Debug, Clone, Copy, PartialEq)]
pub enum SupportedLocale {
    #[serde(rename = "ar-SA")]
    ArSA,
    #[serde(rename = "de-DE")]
    DeDE,
    #[serde(rename = "en-GB")]
    EnGB,
    #[serde(rename = "es-ES")]
    EsES,
    #[serde(rename = "fr-FR")]
    FrFR,
    #[serde(rename = "he-IL")]
    HeIL,
    #[serde(rename = "it-IT")]
    ItIT,
    #[serde(rename = "ja-JP")]
    JaJP,
    #[serde(rename = "ko-KR")]
    KoKR,
    #[serde(rename = "nb-NO")]
    NbNO,
    #[serde(rename = "nl-NL")]
    NlNL,
    #[serde(rename = "pl-PL")]
    PlPL,
    #[serde(rename = "pt-BR")]
    PtBR,
    #[serde(rename = "sq-AL")]
    SqAL,
    #[serde(rename = "sv-SE")]
    SvSE,
    #[serde(rename = "tr-TR")]
    TrTR,
    #[serde(rename = "vi-VN")]
    ViVN,
    #[serde(rename = "zh-CN")]
    ZhCN,
}
impl_optional_overwrite!(SupportedLocale);

//TODO? macro to reduce duplication?
impl SupportedLocale {
    pub fn to_str(self) -> &'static str {
        match self {
            SupportedLocale::ArSA => "ar-SA",
            SupportedLocale::DeDE => "de-DE",
            SupportedLocale::EnGB => "en-GB",
            SupportedLocale::EsES => "es-ES",
            SupportedLocale::FrFR => "fr-FR",
            SupportedLocale::HeIL => "he-IL",
            SupportedLocale::ItIT => "it-IT",
            SupportedLocale::JaJP => "ja-JP",
            SupportedLocale::KoKR => "ko-KR",
            SupportedLocale::NbNO => "nb-NO",
            SupportedLocale::NlNL => "nl-NL",
            SupportedLocale::PlPL => "pl-PL",
            SupportedLocale::PtBR => "pt-BR",
            SupportedLocale::SqAL => "sq-AL",
            SupportedLocale::SvSE => "sv-SE",
            SupportedLocale::TrTR => "tr-TR",
            SupportedLocale::ViVN => "vi-VN",
            SupportedLocale::ZhCN => "zg-CN",
        }
    }
}
