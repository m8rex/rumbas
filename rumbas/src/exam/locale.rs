use rumbas_support::preamble::*;
use schemars::JsonSchema;
use serde::Deserialize;
use serde::Serialize;
use comparable::Comparable;

#[derive(Input, Overwrite, RumbasCheck, Examples)]
#[input(name = "LocaleInput")]
#[derive(Serialize, Deserialize, Comparable, Debug, Clone, JsonSchema, PartialEq)]
pub struct Locale {
    pub name: String, //TODO: document names best used for shareability?
    /// The locale to use in the Numbas interface
    pub numbas_locale: SupportedLocale,
}

macro_rules! create_support_locale {
    ($($name: ident => $key: literal),*) => {
        #[derive(Input, Overwrite, RumbasCheck, Examples)]
        #[input(name = "SupportedLocaleInput")]
        /// Locales supported by Numbas
        #[derive(Serialize, Deserialize, Comparable, JsonSchema, Debug, Clone, Copy, PartialEq)]
        pub enum SupportedLocale {
            $(
                #[serde(rename = $key)]
                $name
            ),*
        }

        //TODO? macro to reduce duplication?
        impl SupportedLocale {
            pub fn to_str(self) -> &'static str {
                match self {
                    $(SupportedLocale::$name => $key),*
                }
            }
        }

    }
}

create_support_locale! {
    ArSA => "ar-SA",
    DeDE => "de-DE",
    EnGB => "en-GB",
    EsES => "es-ES",
    FrFR => "fr-FR",
    HeIL => "he-IL",
    ItIT => "it-IT",
    JaJP => "ja-JP",
    KoKR => "ko-KR",
    NbNO => "nb-NO",
    NlNL => "nl-NL",
    PlPL => "pl-PL",
    PtBR => "pt-BR",
    SqAL => "sq-AL",
    SvSE => "sv-SE",
    TrTR => "tr-TR",
    ViVN => "vi-VN",
    ZhCN => "zg-CN"
}
