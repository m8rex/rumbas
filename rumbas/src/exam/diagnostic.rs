use crate::exam::feedback::Feedback;
use crate::exam::locale::SupportedLocale;
use crate::exam::locale::{Locale, Locales};
use crate::exam::navigation::DiagnosticNavigation;
use crate::exam::numbas_settings::NumbasSettings;
use crate::exam::question_group::QuestionGroups;
use crate::exam::question_group::QuestionPath;
use crate::exam::timing::Timing;
use crate::question::custom_part_type::CustomPartTypeDefinitionPath;
use crate::question::extension::Extensions;
use crate::support::to_numbas::ToNumbas;
use crate::support::to_rumbas::ToRumbas;
use crate::support::translatable::JMENotesTranslatableString;
use crate::support::translatable::TranslatableString;
use crate::support::translatable::TranslatableStrings;
use rumbas_support::preamble::*;
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[derive(Input, Overwrite, RumbasCheck)]
#[input(name = "DiagnosticExamInput")]
#[derive(Serialize, Deserialize, Debug, Clone, JsonSchema)]
/// A Diagnostic Exam
pub struct DiagnosticExam {
    /// All locales for which the exam should be generated
    pub locales: Locales,
    /// The name of the exam
    pub name: TranslatableString,
    /// The navigation settings for this exam
    pub navigation: DiagnosticNavigation,
    /// The timing settings for this exam
    pub timing: Timing,
    /// The feedback settings for this exam
    pub feedback: Feedback,
    /// The questions groups for this exam
    pub question_groups: QuestionGroups,
    /// The settings to set for numbas
    pub numbas_settings: NumbasSettings,
    /// The diagnostic data
    pub diagnostic: Diagnostic,
}

impl ToNumbas<numbas::exam::exam::Exam> for DiagnosticExam {
    fn to_numbas(&self, locale: &str) -> numbas::exam::exam::Exam {
        let basic_settings = self.to_numbas(locale);

        let navigation = self.navigation.to_numbas(locale);

        let timing = self.timing.to_numbas(locale);

        let feedback = self.feedback.to_numbas(locale);

        //TODO
        let functions = Some(HashMap::new());

        //TODO
        let variables = Some(HashMap::new());

        let question_groups: Vec<numbas::exam::question_group::QuestionGroup> =
            self.question_groups.to_numbas(locale);

        let resources: Vec<numbas::question::resource::Resource> = self
            .question_groups
            .iter()
            .flat_map(|qg| {
                qg.clone()
                    .questions
                    .into_iter()
                    .flat_map(|q| q.question_data.resources)
            })
            .collect::<std::collections::HashSet<_>>()
            .into_iter()
            .collect::<Vec<_>>()
            .to_numbas(locale); // TODO: extract?

        let extensions: Vec<String> = self
            .question_groups
            .iter()
            .flat_map(|qg| {
                qg.clone()
                    .questions
                    .into_iter()
                    .map(|q| q.question_data.extensions) // todo: extract?
            })
            .fold(Extensions::default(), Extensions::combine)
            .to_paths();

        let custom_part_types: Vec<numbas::question::custom_part_type::CustomPartType> = self
            .question_groups
            .iter()
            .flat_map(|qg| {
                qg.clone()
                    .questions
                    .into_iter()
                    .flat_map(|q| q.question_data.custom_part_types)
            })
            .collect::<std::collections::HashSet<_>>()
            .into_iter()
            .collect::<Vec<_>>()
            .to_numbas(locale); // todo extract?

        let diagnostic = Some(self.diagnostic.to_numbas(locale));

        numbas::exam::exam::Exam {
            basic_settings,
            resources,
            extensions,
            custom_part_types,
            navigation,
            timing,
            feedback,
            functions,
            variables,
            question_groups,
            diagnostic,
        }
    }
}

impl ToNumbas<numbas::exam::exam::BasicExamSettings> for DiagnosticExam {
    fn to_numbas(&self, locale: &str) -> numbas::exam::exam::BasicExamSettings {
        numbas::exam::exam::BasicExamSettings {
            name: self.name.to_numbas(locale),
            duration_in_seconds: self.timing.duration_in_seconds.to_numbas(locale),
            percentage_needed_to_pass: self.feedback.percentage_needed_to_pass.to_numbas(locale),
            show_question_group_names: Some(
                self.navigation.shared_data.show_names_of_question_groups,
            ),
            show_student_name: Some(self.feedback.clone().show_name_of_student),
            allow_printing: Some(self.navigation.shared_data.allow_printing),
        }
    }
}

#[derive(Input, Overwrite, RumbasCheck)]
#[input(name = "DiagnosticInput")]
#[derive(Serialize, Deserialize, Debug, Clone, JsonSchema)]
/// Information needed for a diagnostic test
pub struct Diagnostic {
    /// The script to use
    script: DiagnosticScript,
    /// The learning objectives,
    objectives: LearningObjectives,
    /// The learning topics
    topics: LearningTopics,
}

impl ToNumbas<numbas::exam::diagnostic::Diagnostic> for Diagnostic {
    fn to_numbas(&self, locale: &str) -> numbas::exam::diagnostic::Diagnostic {
        numbas::exam::diagnostic::Diagnostic {
            knowledge_graph: self.to_numbas(locale),
            script: self.script.to_numbas(locale),
            custom_script: self.script.to_numbas(locale),
        }
    }
}

impl ToNumbas<numbas::exam::diagnostic::DiagnosticKnowledgeGraph> for Diagnostic {
    fn to_numbas(&self, locale: &str) -> numbas::exam::diagnostic::DiagnosticKnowledgeGraph {
        numbas::exam::diagnostic::DiagnosticKnowledgeGraph {
            topics: self.topics.to_numbas(locale),
            learning_objectives: self.objectives.to_numbas(locale),
        }
    }
}

impl ToRumbas<Diagnostic> for numbas::exam::diagnostic::Diagnostic {
    fn to_rumbas(&self) -> Diagnostic {
        Diagnostic {
            script: self.to_rumbas(),
            objectives: self.knowledge_graph.learning_objectives.to_rumbas(),
            topics: self.knowledge_graph.topics.to_rumbas(),
        }
    }
}

#[derive(Input, Overwrite, RumbasCheck)]
#[input(name = "DiagnosticScriptInput")]
#[derive(Serialize, Deserialize, Debug, Clone, JsonSchema, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum DiagnosticScript {
    Mastery,
    Diagnosys,
    Custom(JMENotesTranslatableString),
}

impl ToNumbas<numbas::exam::diagnostic::DiagnosticScript> for DiagnosticScript {
    fn to_numbas(&self, _locale: &str) -> numbas::exam::diagnostic::DiagnosticScript {
        match self {
            DiagnosticScript::Mastery => numbas::exam::diagnostic::DiagnosticScript::Mastery,
            DiagnosticScript::Custom(_) => numbas::exam::diagnostic::DiagnosticScript::Custom,
            DiagnosticScript::Diagnosys => numbas::exam::diagnostic::DiagnosticScript::Diagnosys,
        }
    }
}

impl ToNumbas<numbas::jme::JMENotesString> for DiagnosticScript {
    fn to_numbas(&self, locale: &str) -> numbas::jme::JMENotesString {
        match self {
            DiagnosticScript::Custom(s) => s.to_numbas(locale),
            DiagnosticScript::Diagnosys => Default::default(),
            DiagnosticScript::Mastery => Default::default(),
        }
    }
}

impl ToRumbas<DiagnosticScript> for numbas::exam::diagnostic::Diagnostic {
    fn to_rumbas(&self) -> DiagnosticScript {
        match self.script {
            numbas::exam::diagnostic::DiagnosticScript::Mastery => DiagnosticScript::Mastery,
            numbas::exam::diagnostic::DiagnosticScript::Diagnosys => DiagnosticScript::Diagnosys,
            numbas::exam::diagnostic::DiagnosticScript::Custom => {
                DiagnosticScript::Custom(self.custom_script.clone().into())
            }
        }
    }
}

pub type LearningObjectivesInput = Vec<Value<LearningObjectiveInput>>;
pub type LearningObjectives = Vec<LearningObjective>;

#[derive(Input, Overwrite, RumbasCheck)]
#[input(name = "LearningObjectiveInput")]
#[derive(Serialize, Deserialize, Debug, Clone, JsonSchema, PartialEq)]
/// A Learning Objective
pub struct LearningObjective {
    /// The name
    name: TranslatableString,
    /// A description
    description: TranslatableString,
}

impl ToNumbas<numbas::exam::diagnostic::DiagnosticKnowledgeGraphLearningObjective>
    for LearningObjective
{
    fn to_numbas(
        &self,
        locale: &str,
    ) -> numbas::exam::diagnostic::DiagnosticKnowledgeGraphLearningObjective {
        numbas::exam::diagnostic::DiagnosticKnowledgeGraphLearningObjective {
            name: self.name.to_numbas(locale),
            description: self.description.to_numbas(locale),
        }
    }
}

impl ToRumbas<LearningObjective>
    for numbas::exam::diagnostic::DiagnosticKnowledgeGraphLearningObjective
{
    fn to_rumbas(&self) -> LearningObjective {
        LearningObjective {
            name: self.name.to_rumbas(),
            description: self.description.to_rumbas(),
        }
    }
}

pub type LearningTopicsInput = Vec<Value<LearningTopicInput>>;
pub type LearningTopics = Vec<LearningTopic>;

#[derive(Input, Overwrite, RumbasCheck)]
#[input(name = "LearningTopicInput")]
#[derive(Serialize, Deserialize, Debug, Clone, JsonSchema, PartialEq)]
/// A learning Topic
pub struct LearningTopic {
    /// The name
    name: TranslatableString,
    /// A description
    description: TranslatableString,
    /// List of names of objectives
    objectives: TranslatableStrings,
    /// List of names of topic on which this topic depends
    depends_on: TranslatableStrings,
}

impl ToNumbas<numbas::exam::diagnostic::DiagnosticKnowledgeGraphTopic> for LearningTopic {
    fn to_numbas(&self, locale: &str) -> numbas::exam::diagnostic::DiagnosticKnowledgeGraphTopic {
        numbas::exam::diagnostic::DiagnosticKnowledgeGraphTopic {
            name: self.name.to_numbas(locale),
            description: self.description.to_numbas(locale),
            learning_objectives: self.objectives.to_numbas(locale),
            depends_on: self.depends_on.to_numbas(locale),
        }
    }
}

impl ToRumbas<LearningTopic> for numbas::exam::diagnostic::DiagnosticKnowledgeGraphTopic {
    fn to_rumbas(&self) -> LearningTopic {
        LearningTopic {
            name: self.name.to_rumbas(),
            description: self.description.to_rumbas(),
            objectives: self.learning_objectives.to_rumbas(),
            depends_on: self.depends_on.to_rumbas(),
        }
    }
}

/// Converts a diagnostic numbas exam to a DiagnosticExam and extracts questions and
/// custom_part_types
pub fn convert_diagnostic_numbas_exam(
    exam: numbas::exam::exam::Exam,
) -> (
    DiagnosticExam,
    Vec<QuestionPath>,
    Vec<CustomPartTypeDefinitionPath>,
) {
    let question_groups: Vec<_> = exam.question_groups.to_rumbas();
    let custom_part_types = exam.custom_part_types.to_rumbas();
    (
        DiagnosticExam {
            locales: vec![Locale {
                name: "en".to_string(),
                numbas_locale: SupportedLocale::EnGB,
            }], // todo: argument?
            name: exam.basic_settings.name.to_rumbas(),
            navigation: exam.to_rumbas(),
            timing: exam.to_rumbas(),
            feedback: exam.to_rumbas(),
            question_groups: question_groups.clone(),
            numbas_settings: NumbasSettings {
                theme: "default".to_string(),
            }, // todo: argument?
            diagnostic: exam.diagnostic.unwrap().to_rumbas(), // Always set for a diagnostic exam
        },
        question_groups
            .into_iter()
            .flat_map(|qg| qg.questions)
            .collect(),
        custom_part_types,
    )
}
