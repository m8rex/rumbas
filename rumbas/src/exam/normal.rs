use crate::exam::feedback::Feedback;
use crate::exam::locale::Locale;
use crate::exam::locale::SupportedLocale;
use crate::exam::navigation::NormalNavigation;
use crate::exam::numbas_settings::NumbasSettings;
use crate::exam::question_group::QuestionGroup;
use crate::exam::question_group::QuestionPath;
use crate::exam::timing::Timing;
use crate::question::custom_part_type::CustomPartTypeDefinitionPath;
use crate::question::extension::Extensions;
use crate::support::to_numbas::ToNumbas;
use crate::support::to_rumbas::ToRumbas;
use crate::support::translatable::TranslatableString;
use rumbas_support::preamble::*;
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use serde_diff::{Apply, Diff, SerdeDiff};
use std::collections::HashMap;

// TODO: remove duplication of NormalExam & Diagnostic Exam?
#[derive(Input, Overwrite, RumbasCheck, Examples)]
#[input(name = "NormalExamInput")]
#[derive(Serialize, Deserialize, SerdeDiff, Debug, Clone, JsonSchema, PartialEq)]
/// An Exam
pub struct NormalExam {
    /// All locales for which the exam should be generated
    pub locales: Vec<Locale>,
    /// The name of the exam
    pub name: TranslatableString,
    /// The navigation settings for this exam
    pub navigation: NormalNavigation,
    /// The timing settings for this exam
    pub timing: Timing,
    /// The feedback settings for this exam
    pub feedback: Feedback,
    /// The questions groups for this exam
    pub question_groups: Vec<QuestionGroup>,
    /// The settings to set for numbas
    pub numbas_settings: NumbasSettings,
}

impl ToNumbas<numbas::exam::Exam> for NormalExam {
    fn to_numbas(&self, locale: &str) -> numbas::exam::Exam {
        let basic_settings = self.to_numbas(locale);
        let navigation = self.navigation.to_numbas(locale);

        let timing = self.timing.to_numbas(locale);

        let feedback = self.feedback.to_numbas(locale);

        //TODO
        let functions = Some(HashMap::new());

        //TODO
        let variables = Some(HashMap::new());

        let question_groups: Vec<numbas::exam::question_group::QuestionGroup> =
            self.question_groups.to_numbas(locale);

        let resources: Vec<numbas::question::resource::Resource> = self
            .question_groups
            .iter()
            .flat_map(|qg| {
                qg.clone()
                    .questions
                    .into_iter()
                    .flat_map(|q| q.data.resources)
            })
            .collect::<std::collections::HashSet<_>>()
            .into_iter()
            .collect::<Vec<_>>()
            .to_numbas(locale);

        let extensions: Vec<String> = self
            .question_groups
            .iter()
            .flat_map(|qg| qg.clone().questions.into_iter().map(|q| q.data.extensions))
            .fold(Extensions::default(), Extensions::combine)
            .to_paths();

        let custom_part_types: Vec<numbas::question::custom_part_type::CustomPartType> = self
            .question_groups
            .clone()
            .iter()
            .flat_map(|qg| {
                qg.clone()
                    .questions
                    .into_iter()
                    .flat_map(|q| q.data.custom_part_types)
            })
            .collect::<std::collections::HashSet<_>>()
            .into_iter()
            .collect::<Vec<_>>()
            .to_numbas(locale);

        numbas::exam::Exam {
            basic_settings,
            resources,
            extensions,
            custom_part_types,
            navigation,
            timing,
            feedback,
            functions,
            variables,
            question_groups,
            diagnostic: None,
        }
    }
}

impl ToNumbas<numbas::exam::BasicExamSettings> for NormalExam {
    fn to_numbas(&self, locale: &str) -> numbas::exam::BasicExamSettings {
        numbas::exam::BasicExamSettings {
            name: self.name.to_numbas(locale),
            duration_in_seconds: self.timing.duration_in_seconds.to_numbas(locale),
            percentage_needed_to_pass: self.feedback.percentage_needed_to_pass.to_numbas(locale),
            show_question_group_names: Some(
                self.navigation
                    .to_shared_data()
                    .show_names_of_question_groups,
            ),
            show_student_name: Some(self.feedback.show_name_of_student),
            allow_printing: Some(self.navigation.to_shared_data().allow_printing),
        }
    }
}

/// Converts a normal numbas exam to a NormalExam
pub fn convert_normal_numbas_exam(
    exam: numbas::exam::Exam,
) -> (
    NormalExam,
    Vec<QuestionPath>,
    Vec<CustomPartTypeDefinitionPath>,
) {
    let question_groups: Vec<_> = exam.question_groups.to_rumbas();
    let custom_part_types = exam.custom_part_types.to_rumbas();
    (
        NormalExam {
            locales: vec![Locale {
                name: "en".to_string(),
                numbas_locale: SupportedLocale::EnGB,
            }], // todo: argument?
            name: exam.basic_settings.name.to_rumbas(),
            navigation: exam.to_rumbas(),
            timing: exam.to_rumbas(),
            feedback: exam.to_rumbas(),
            question_groups: question_groups.clone(),
            numbas_settings: NumbasSettings {
                theme: "default".to_string(),
            }, // todo: argument?
        },
        question_groups
            .into_iter()
            .flat_map(|qg| qg.questions.into_iter())
            .collect(),
        custom_part_types,
    )
}
