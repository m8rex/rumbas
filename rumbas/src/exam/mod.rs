//! Contains all the exam types

pub mod diagnostic;
pub mod feedback;
pub mod locale;
pub mod navigation;
pub mod normal;
pub mod numbas_settings;
pub mod question_group;
pub mod timing;

use crate::exam::diagnostic::convert_diagnostic_numbas_exam;
use crate::exam::diagnostic::DiagnosticExam;
use crate::exam::locale::Locale;
use crate::exam::normal::convert_normal_numbas_exam;
use crate::exam::normal::NormalExam;
use crate::exam::question_group::QuestionPath;
use crate::question::custom_part_type::CustomPartTypeDefinitionPath;
use crate::support::template::{TemplateFile, TemplateFileInputEnum};
use crate::support::to_numbas::ToNumbas;
use crate::support::translatable::TranslatableString;
use crate::support::yaml::YamlError;
use rumbas_support::preamble::*;
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::fmt::Display;
use std::fs;
use std::path::Path;

#[derive(Input, Overwrite, RumbasCheck, Examples)]
#[input(name = "ExamInput")]
#[input(test)]
#[derive(Serialize, Deserialize, Debug, Clone, JsonSchema, PartialEq)]
#[serde(rename_all = "snake_case")]
#[serde(tag = "type")]
pub enum Exam {
    Normal(NormalExam),
    Diagnostic(DiagnosticExam),
}

impl ToNumbas<numbas::exam::Exam> for Exam {
    fn to_numbas(&self, locale: &str) -> numbas::exam::Exam {
        match self {
            Exam::Normal(n) => n.to_numbas(locale),
            Exam::Diagnostic(n) => n.to_numbas(locale),
        }
    }
}

impl Exam {
    pub fn locales(&self) -> Vec<Locale> {
        match self {
            Exam::Normal(n) => n.locales.clone(),
            Exam::Diagnostic(n) => n.locales.clone(),
        }
    }

    pub fn numbas_settings(&self) -> crate::exam::numbas_settings::NumbasSettings {
        match self {
            Exam::Normal(n) => n.numbas_settings.clone(),
            Exam::Diagnostic(n) => n.numbas_settings.clone(),
        }
    }
}
impl ExamInput {
    pub fn from_file(file: &Path) -> Result<ExamInput, ParseError> {
        use ExamFileTypeInput::*;
        let input: std::result::Result<ExamFileTypeInput, serde_yaml::Error> =
            if file.starts_with(crate::EXAMS_FOLDER) {
                let yaml = fs::read_to_string(file).map_err(ParseError::IOError)?;
                serde_yaml::from_str(&yaml)
            } else if file.starts_with(crate::QUESTIONS_FOLDER) {
                let mut data = HashMap::new();
                data.insert(
                    "question".to_string(),
                    serde_yaml::Value::String(
                        file.with_extension("")
                            .strip_prefix(crate::QUESTIONS_FOLDER)
                            .unwrap()
                            .to_string_lossy()
                            .into_owned(),
                    )
                    .into(),
                );
                let t = TemplateFile {
                    relative_template_path: crate::QUESTION_PREVIEW_TEMPLATE_NAME.to_string(),
                    data,
                };
                Ok(Template(TemplateFileInputEnum::from_normal(t)))
            } else {
                panic!(
                    "{} should start with {}/ or {}/",
                    file.display(),
                    crate::EXAMS_FOLDER,
                    crate::QUESTIONS_FOLDER
                );
            };
        input
            .map(|e| match e {
                Normal(e) => Ok(ExamInput::Normal(e)),
                Diagnostic(e) => Ok(ExamInput::Diagnostic(e)),
                Template(t_val) => {
                    let t = t_val.to_normal();
                    let template_file = Path::new(crate::EXAM_TEMPLATES_FOLDER)
                        .join(format!("{}.yaml", t.relative_template_path)); // TODO: check for missing fields.....
                    let template_yaml = fs::read_to_string(&template_file).expect(
                        &format!(
                            "Failed to read {}",
                            template_file.to_str().map_or("invalid filename", |s| s)
                        )[..],
                    );

                    let mut exam: ExamInput = serde_yaml::from_str(&template_yaml).unwrap();
                    t.data.iter().for_each(|(k, v)| {
                        exam.insert_template_value(k, &v.0);
                    });
                    Ok(exam)
                }
            })
            .and_then(std::convert::identity) //flatten result is currently only possible in nightly
            .map_err(|e| ParseError::YamlError(YamlError::from(e, file.to_path_buf())))
    }
}

#[derive(Debug, Display)]
pub enum ParseError {
    YamlError(YamlError),
    IOError(std::io::Error),
}

#[derive(Input, Overwrite, RumbasCheck)]
#[input(name = "ExamFileTypeInput")]
#[derive(Serialize, Deserialize, Debug, Clone, JsonSchema)]
#[serde(rename_all = "snake_case")]
#[serde(tag = "type")]
pub enum ExamFileType {
    Template(TemplateFile),
    Normal(NormalExam),
    Diagnostic(DiagnosticExam),
}

impl ExamFileTypeInput {
    pub fn to_yaml(&self) -> serde_yaml::Result<String> {
        serde_yaml::to_string(self)
    }
}

impl ExamFileType {
    pub fn to_yaml(&self) -> serde_yaml::Result<String> {
        ExamFileTypeInput::from_normal(self.to_owned()).to_yaml()
    }
}

/// Convert a numbas exam to rumbas data
/// Returns the name of the exam, the resulting exam (as ExamFileType)
/// and vectors of questions and custom part type definitions
pub fn convert_numbas_exam(
    exam: numbas::exam::Exam,
) -> (
    String,
    ExamFileType,
    Vec<QuestionPath>,
    Vec<CustomPartTypeDefinitionPath>,
) {
    let (name, exam, qgs, cpts) = match exam.navigation.navigation_mode {
        numbas::exam::navigation::NavigationMode::Diagnostic(ref _d) => {
            let (exam, qgs, cpts) = convert_diagnostic_numbas_exam(exam);
            (exam.name.clone(), ExamFileType::Diagnostic(exam), qgs, cpts)
        }
        _ => {
            let (exam, qgs, cpts) = convert_normal_numbas_exam(exam);
            (exam.name.clone(), ExamFileType::Normal(exam), qgs, cpts)
        }
    };
    (
        {
            if let TranslatableString::NotTranslated(n) = name {
                n.get_content("").unwrap()
            } else {
                panic!("Should not happen");
            }
        },
        exam,
        qgs,
        cpts,
    )
}
