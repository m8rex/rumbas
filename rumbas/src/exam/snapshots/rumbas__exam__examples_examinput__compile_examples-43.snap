---
source: rumbas/src/exam/mod.rs
expression: "&example"
---
type: diagnostic
locales:
  - name: nonjmetext§
    numbas_locale: ar-SA
  - name: "template:template_key"
    numbas_locale: de-DE
  - name:
      template_key: template_key
      default_value: ~
    numbas_locale: en-GB
  - name: ~
    numbas_locale: es-ES
  - name: nonjmetext§
    numbas_locale: fr-FR
  - name: "template:template_key"
    numbas_locale: he-IL
  - name:
      template_key: template_key
      default_value: ~
    numbas_locale: it-IT
  - name: ~
    numbas_locale: ja-JP
  - name: nonjmetext§
    numbas_locale: ko-KR
  - name: "template:template_key"
    numbas_locale: nb-NO
  - name:
      template_key: template_key
      default_value: ~
    numbas_locale: nl-NL
  - name: ~
    numbas_locale: pl-PL
  - name: nonjmetext§
    numbas_locale: pt-BR
  - name: "template:template_key"
    numbas_locale: sq-AL
  - name:
      template_key: template_key
      default_value: ~
    numbas_locale: sv-SE
  - name: ~
    numbas_locale: tr-TR
  - name: nonjmetext§
    numbas_locale: vi-VN
  - name: "template:template_key"
    numbas_locale: zg-CN
  - name:
      template_key: template_key
      default_value: ~
    numbas_locale: "template:template_key"
  - name: ~
    numbas_locale:
      template_key: template_key
      default_value: ~
  - name: nonjmetext§
    numbas_locale: ~
  - "template:template_key"
  - template_key: template_key
    default_value: ~
name:
  content:
    nonjmetext§: "file:path/to/file"
  placeholders:
    placeholder1:
      content:
        en: en value of placeholder1
        nl: nl value of placeholder1
      placeholders: {}
    placeholder2:
      content:
        en: en version of placeholder2
        nl: nl version of placeholder2
      placeholders: {}
navigation:
  allow_printing: "template:template_key"
  can_regenerate: "template:template_key"
  confirm_when_leaving: "template:template_key"
  on_leave:
    template_key: template_key
    default_value: ~
  show_names_of_question_groups: "template:template_key"
  show_steps: "template:template_key"
  show_title_page: "template:template_key"
  start_password: none
timing:
  duration_in_seconds: "template:template_key"
  allow_pause:
    template_key: template_key
    default_value: ~
  on_timeout:
    action: warn
    message:
      content:
        nonjmetext§: "template:template_key"
      placeholders:
        placeholder1:
          content:
            en: en value of placeholder1
            nl: nl value of placeholder1
          placeholders: {}
        placeholder2:
          content:
            en: en version of placeholder2
            nl: nl version of placeholder2
          placeholders: {}
  timed_warning:
    action: warn
    message:
      content:
        nonjmetext§: "template:template_key"
      placeholders:
        placeholder1:
          content:
            en: en value of placeholder1
            nl: nl value of placeholder1
          placeholders: {}
        placeholder2:
          content:
            en: en version of placeholder2
            nl: nl version of placeholder2
          placeholders: {}
feedback:
  percentage_needed_to_pass: 1.2
  show_name_of_student: "template:template_key"
  show_current_marks: "template:template_key"
  show_maximum_marks: "template:template_key"
  show_answer_state: "template:template_key"
  allow_reveal_answer: "template:template_key"
  review:
    template_key: template_key
    default_value: ~
  advice: "template:template_key"
  intro: "template:template_key"
  feedback_messages: "template:template_key"
question_groups:
  - name:
      content:
        nonjmetext§: "file:path/to/file"
      placeholders:
        placeholder1:
          content:
            en: en value of placeholder1
            nl: nl value of placeholder1
          placeholders: {}
        placeholder2:
          content:
            en: en version of placeholder2
            nl: nl version of placeholder2
          placeholders: {}
    picking_strategy: all_ordered
    questions:
      - path
      - "template:template_key"
      - template_key: template_key
        default_value: ~
  - name:
      content:
        nonjmetext§: "template:template_key"
      placeholders:
        placeholder1:
          content:
            en: en value of placeholder1
            nl: nl value of placeholder1
          placeholders: {}
        placeholder2:
          content:
            en: en version of placeholder2
            nl: nl version of placeholder2
          placeholders: {}
    picking_strategy: all_shuffled
    questions: "template:template_key"
  - name:
      content:
        nonjmetext§:
          template_key: template_key
          default_value: ~
      placeholders:
        placeholder1:
          content:
            en: en value of placeholder1
            nl: nl value of placeholder1
          placeholders: {}
        placeholder2:
          content:
            en: en version of placeholder2
            nl: nl version of placeholder2
          placeholders: {}
    pick_questions: 2
    picking_strategy: random_subset
    questions:
      template_key: template_key
      default_value: ~
  - name:
      content: "file:path/to/file"
      placeholders:
        placeholder1:
          content:
            en: en value of placeholder1
            nl: nl value of placeholder1
          placeholders: {}
        placeholder2:
          content:
            en: en version of placeholder2
            nl: nl version of placeholder2
          placeholders: {}
    pick_questions: "template:template_key"
    picking_strategy: random_subset
    questions: ~
  - name: "file:path/to/file"
    pick_questions:
      template_key: template_key
      default_value: ~
    picking_strategy: random_subset
    questions:
      - path
      - "template:template_key"
      - template_key: template_key
        default_value: ~
  - name: "template:template_key"
    picking_strategy: all_ordered
    questions: "template:template_key"
  - name:
      template_key: template_key
      default_value: ~
    picking_strategy: all_shuffled
    questions:
      template_key: template_key
      default_value: ~
  - name: ~
    pick_questions: 2
    picking_strategy: random_subset
    questions: ~
  - "template:template_key"
  - template_key: template_key
    default_value: ~
numbas_settings:
  theme:
    template_key: template_key
    default_value: ~
diagnostic:
  script:
    custom:
      content:
        nonjmetext§: "template:template_key"
      placeholders:
        placeholder1:
          content:
            en: en value of placeholder1
            nl: nl value of placeholder1
          placeholders: {}
        placeholder2:
          content:
            en: en version of placeholder2
            nl: nl version of placeholder2
          placeholders: {}
  objectives: ~
  topics: ~

